### PR TITLE
[web] Fix paths fetched by flutter.js

### LINKED ISSUE #116360: [web] PathUrlStrategy doesn't work with nested paths
## Steps to Reproduce

Channel master

1. Set `PathUrlStrategy` on default project
2. Run locally and try to navigate to /a/b

Error: `The script has an unsupported MIME type ('text/html').`

But if you change entry point script in `index.html` from
```
<script>
  window.addEventListener('load', function(ev) {
    // Download main.dart.js
    _flutter.loader.loadEntrypoint({
      serviceWorker: {
        serviceWorkerVersion: serviceWorkerVersion,
      },
      onEntrypointLoaded: function(engineInitializer) {
        engineInitializer.initializeEngine().then(function(appRunner) {
          appRunner.runApp();
        });
      }
    });
  });
</script>
```

to
```
<script src="main.dart.js" type="application/javascript"></script>
```

then it works as expected.

### PR DESCRIPTION
**Pull request by mdebbar**
### This fix may be cherry-picked into 3.7.X

This PR is in the process of being cherry-picked for the current `stable` channel (3.7.?). Click the link to [follow the progress of the cherry-pick](https://github.com/flutter/flutter/issues/120197).

### Workaround

In the meantime, you can also configure the main.dart.js and flutter_service_worker.js URLs as described in this comment to something that matches your app:

* https://github.com/flutter/flutter/issues/116360#issuecomment-1364181861

TL;DR:

```js
      _flutter.loader.loadEntrypoint({
        entrypointUrl: "/YOUR_APP_BASE/main.dart.js",
        serviceWorker: {
          serviceWorkerVersion: serviceWorkerVersion,
          serviceWorkerUrl: "/YOUR_APP_BASE/flutter_service_worker.js?v=",
        },
   ...
```

That way the files will be requested from their "absolute" URL, rather than URLs relative to the ones specified by the PathStrategy.

(If your base URL is "/", you just need: `/main.dart.js` and `/flutter_service_worker.js?v=`)

----

`flutter.js` fetches `main.dart.js` and `flutter_service_worker.js` using relative paths. That works in the default case of most flutter web apps, but when an app uses `<base href="/foo/">` or use the `PathUrlStrategy`, things break.

This PR adjusts all requests made by `flutter.js` to take `<base href="...">` into account.

Fixes https://github.com/flutter/flutter/issues/116360

### CODE REVIEW

**Comment by mdebbar** [comment]
I extracted this PR out of https://github.com/flutter/flutter/pull/117691 because it looks like that one will take a while to land. So let's fix the actual issue now, and worry about improving the ergonomics of generated files later.

File: packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart [hunk 1]
```diff
@@ -20,6 +20,21 @@ _flutter.loader = null;
 
 (function () {
   "use strict";
+
+  const baseUri = ensureTrailingSlash(getBaseURI());
+
+  function getBaseURI() {
+    const base = document.querySelector("base");
+    return (base && base.getAttribute("href")) || "";
+  }
```
Comment (ditman) [inline packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart:29]: Would it be possible to use `document.baseURI` to retrieve this instead of the selector?

https://developer.mozilla.org/en-US/docs/Web/API/Node/baseURI

(It also seems to guarantee the trailing slash?)

**Review by ditman — APPROVED** [review]
Tiny suggestion.

File: packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart (hunk 1 above)
Comment (mdebbar) [reply packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart:29]: That's how I [did it initially](https://github.com/flutter/flutter/pull/117691/commits/4a94af02ce10cb02a3bc7c25d29b54e8a655d3e3#diff-818be2e3cd2b55c69070b0fca9216af012cc8e42ba30f02bfcad697dcdb3c71cR13) but that was causing mysterious test failures (the test harness launches Chrome and then waits forever but Chrome never connects back). Turned out that was because some tests (specifically integration tests), don't have a `<base href="...">` and the test is served on a sub-path that's not the root, but `document.baseURI` points to `/` so it makes all the paths become absolute paths.

I had to remove `document.baseURI` and instead get the `href` from the `<base>` element ([here](https://github.com/flutter/flutter/pull/117691/commits/f0e0e35bf1742e69b20f7995c17493e32c81557d)). If a `<base>` element isn't available, we default to an empty string, which has no effect on the request paths, so everything works correctly.

Comment (ditman) [reply packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart:29]: That's so odd, I'd expect that if the base is not present, `document.baseURI` defaults to whatever sub-path the page is served from (in an absolute URL), rather than "/" :/

Comment (mdebbar) [reply packages/flutter_tools/lib/src/web/file_generators/flutter_js.dart:29]: > That's so odd, I'd expect that if the base is not present, `document.baseURI` defaults to whatever sub-path the page is served from (in an absolute URL), rather than "/" :/

Actually, I think you are right. I may have misremembered. Either way, the default `document.baseURI` wasn't working, so I had to parse manually and default to empty string to make it work.

**Comment by leynier** [comment]
Hi @ditman @mdebbar quick question, is this fix available on the stable channel? I'm having problems with this https://github.com/flutter/flutter/issues/119492 Thanks in advance.

**Comment by RamiroGarrido** [comment]
Hello friends, ty for your hard work with Flutter....I just wanted to ask if u have already a date of release for this commit on the stable channel? My web project isn't working with flutter 3.7.0 due to this issue (as reported in https://github.com/flutter/flutter/issues/116360 too).

**Comment by Sipioteo** [comment]
> Hello friends, ty for your hard work with Flutter....I just wanted to ask if u have already a date of release for this commit on the stable channel? My web project isn't working with flutter 3.7.0 due to this issue (as reported in #116360 too).

Neither mine

**Comment by ditman** [comment]
@leynier @RamiroGarrido @Sipioteo there has been a stable release very recently (last week), so it's unlikely that this will get to stable until the next release (in about 3 months), unless it goes through the cherry-pick process.

Before we even think of cherry-picking: Have you verified that what's currently in channel `master` fixes your issues?

**Comment by RamiroGarrido** [comment]
Hello friends, ty for your hard work with Flutter....I just wanted to ask if u have already a date of release for this commit on the stable channel? My web project isn't working with flutter 3.7.0 due to this issue (as reported in https://github.com/flutter/flutter/issues/116360 too).

> @leynier @RamiroGarrido @Sipioteo there has been a stable release very recently (last week), so it's unlikely that this will get to stable until the next release (in about 3 months), unless it goes through the cherry-pick process. [Documentation](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#stable).
> 
> Before we even think of cherry-picking: Have you verified that what's currently in channel `master` fixes your issues?

Ty for your response ditman!...I wasn't sure to downgrade to my previous flutter version, but now I am.... to answer your question: I updated my web project from 3.3.10 to 3.7.0 three days ago, so I'm sure this issue it's not fixed yet on the stable channel...Blessings

**Comment by ditman** [comment]
> I wasn't sure to downgrade to my previous flutter version, but now I am...

Have you considered upgrading to `flutter master`, rather than downgrading to 3.3?

**Comment by RamiroGarrido** [comment]
> Have you considered upgrading to `flutter master`, rather than downgrading to 3.3?

I thought about it, but I rather wait since the master channel may have broken code, according to the wiki (https://github.com/flutter/flutter/wiki/Flutter-build-release-channels)...I think is better to avoid a bug on my production environment...ty for your suggestion though...maybe the other people reading this thread might give it a shot and report back to us...

**Comment by Sipioteo** [comment]
> @leynier @RamiroGarrido @Sipioteo there has been a stable release very recently (last week), so it's unlikely that this will get to stable until the next release (in about 3 months), unless it goes through the cherry-pick process. [Documentation](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#stable).
> 
> Before we even think of cherry-picking: Have you verified that what's currently in channel `master` fixes your issues?
> 

Yeah, I tried. But some of the packages are not compatibles with the master version (missing a vsync param into an animated widget, I don't remember which one right now).
I have to use an older version of flutter right now, so if it's possible to cherry pick it would be a really great move.

Thank you!

**Comment by Sipioteo** [comment]
I have found a temporary solution.
Put the new flutter.js file into your web folder (in the root of the project).
When you will build it will replace the generated one and it will work (I don't think it work well with deferred but you can try.)

Best to all, waiting for the official integration.

**Comment by maRci002** [comment]
**Workaround for every project:**

1. apply changes to `flutter\packages\flutter_tools\lib\src\web\file_generators\flutter_js.dart` and save
2. remove `flutter\bin\cache\flutter_tools.stamp`
3. run `flutter doctor` to Rebuild flutter tool

When new `stable` is available:
run `git reset --hard` in `flutter` folder otherwise it won't find new releases.

**Comment by stx** [comment]
As far as I can tell, the latest Flutter stable release that includes this bug has broken every Flutter web app that uses paths?

This seems like it should qualify for an extremely urgent cherry pick into stable?

For anyone else coming across this, here's a complete fix:

1. Get my working `flutter.js` from master: https://gist.github.com/stx/2b121f94388bc0f06c8a743a18233ac1
2. Put this file at `web/flutter_override.js`
3. In `web/index.html`, change `<script src="flutter.js" defer></script>` to `<script src="/flutter_override.js" defer></script>`

**Comment by ditman** [comment]
@stx yeah, I'll bring this up to the next cherry-pick. (/cc @mdebbar)

**Comment by ditman** [comment]
@stx have you tried to pass the correct entrypointUrl and serviceWorkerUrl to the loadEntrypoint call, as described here?

https://github.com/flutter/flutter/issues/116360#issuecomment-1364181861

I'd say passing `/YOUR_APP_BASE/main.dart.js` and `/YOUR_APP_BASE/flutter_service_worker.js?v=` should fix the issues without needing to change the flutter.js file itself.

**Comment by stx** [comment]
@ditman I didn't see that solution. Yep, works nicely!

**Comment by mdebbar** [comment]
Thanks everyone for the feedback. I filed an issue to cherry-pick this PR into the stable channel: https://github.com/flutter/flutter/issues/120197

**Comment by ditman** [comment]
Updated the description at the top answering the FAQ about this being cherry-picked, and how to work-around this until the fix arrives.

**Comment by s-hammad-a** [comment]
> As far as I can tell, the latest Flutter stable release that includes this bug has broken every Flutter web app that uses paths?
> 
> This seems like it should qualify for an extremely urgent cherry pick into stable?
> 
> For anyone else coming across this, here's a complete fix for the stable branch that will work on CI systems and without having to mess with stable/master switching:
> 
> In `web/index.html`, change this:
> 
> ```
>         _flutter.loader.loadEntrypoint({
>           serviceWorker: {
>             serviceWorkerVersion: serviceWorkerVersion,
>           }
>         })
> ```
> 
> to this:
> 
> ```
>       _flutter.loader.loadEntrypoint({
>         entrypointUrl: "/main.dart.js",
>         serviceWorker: {
>           serviceWorkerVersion: serviceWorkerVersion,
>           serviceWorkerUrl: "/flutter_service_worker.js?v=",
>         },
>       );
> ```
> 
> Taken from [#116360 (comment)](https://github.com/flutter/flutter/issues/116360#issuecomment-1364181861).

It is not working when I am using "PathUrlStrategy()".

**Comment by MirzaCickusic** [comment]
@Frank3K have you figured out how to use your solution within flutter 3.22?

**Comment by cerealexx** [comment]
@Frank3K your code doesn't work if you refresh the page. Flutter still tries to find `main.dart.js` under the path you're currently in instead of looking at the base.

This code works for me even after refreshing the page:
```
<script>
    {{flutter_js}}
    {{flutter_build_config}}
    const serviceWorkerVersion = {{flutter_service_worker_version}};
    
    window.addEventListener('load', () => {
      _flutter.loader.loadEntrypoint({
        entrypointUrl: "/main.dart.js",
        serviceWorker: {
          serviceWorkerVersion: serviceWorkerVersion,
          serviceWorkerUrl: "/flutter_service_worker.js?v=",
        },
        onEntrypointLoaded: async (engineInitializer) => {
          const appRunner = await engineInitializer.initializeEngine({
              hostElement: document.querySelector("#app-wrapper"),
          });
          appRunner.runApp();
        }
      });
    });
  </script>
```
But I haven't been able to replicate the same behavior when migrating from `loadEntryPoint` to `load`.
`entrypointUrl` seems to just have ceased to exist with no alternative 🤷‍♂️

**Comment by Frank3K** [comment]
> Hi @cerealexx @Frank3K. I'm having the same issue mentioned here, but sadly I'm at a loss as to what file I need to be modifying with your supplied code. Could you please tell me which one it is? Thanks, Andy.

This is the `flutter_bootstrap.js` file.

**Comment by andyshephard** [comment]
> @MirzaCickusic For 3.22 this should work:
> 
> ```js
> {{flutter_js}}
> {{flutter_build_config}}
> 
> const serviceWorkerVersion = {{flutter_service_worker_version}};
> 
> let baseUri = document.baseURI;
> if (baseUri.endsWith('/')) {
>     baseUri = baseUri.slice(0, -1);
> }
> 
> _flutter.loader.load({
>     config: {
>       canvasKitBaseUrl: `${baseUri}/canvaskit/`,
>     },
>     serviceWorkerSettings: {
>       serviceWorkerVersion: serviceWorkerVersion,
>       serviceWorkerUrl: `${baseUri}/flutter_service_worker.js?v=${serviceWorkerVersion}`,
>     },
>     onEntrypointLoaded: function (engineInitializer) {
>       engineInitializer.initializeEngine().then(function (appRunner) {
>         appRunner.runApp();
>       });
>     },
> });
> ```

FWIW, this worked for me however I did have to downgrade Flutter from v3.24.0 to v3.22.3
